Simplify Channel constructors, change kwargs to positional args.

NHDaly opened the pull request:
Oops, fixes this error, introduced in #30855:
```julia
WARNING: Method definition (::Type{Base.Channel{T}})(Function) where {T} in module Base at channels.jl:133 overwritten at channels.jl:136.
```

NHDaly committed:
Fix method overwrite warning for Channel constructors

Oops, fixes this error, introduced in e891bca:
WARNING: Method definition (::Type{Base.Channel{T}})(Function) where {T} in module Base at channels.jl:133 overwritten at channels.jl:136.

NHDaly commented on base/channels.jl:
```diff
@@ -129,8 +129,8 @@ function Channel(func::Function; ctype=Any, csize=0, taskref=nothing)
+    return Channel(f, csize=size, ctype=T)
 end
 function Channel{T}(f::Function; csize=0, taskref=nothing) where T
     return Channel(f, csize=csize, ctype=T, taskref=taskref)
```
@JeffBezanson While i'm here, I wonder if it was a mistake to keep the kwargs in this separate constructor.

Ideally, what i would have done is delete the `csize=0` and `ctype=` kwargs, and merge the constructors together so that we just had two types of constructors:
- `Channel{T=Any}(func::Function, size=0; taskref=nothing)`: Creates a task w/ the provided `func`
- `Channel{T=Any}(sz=0)`: Just creates a Channel

But since we can't delete existing params (no breaking changes), we had to just add _more_ options.

Do you think it would make sense to also add the optional `; taskref=nothing` argument to the `Channel{T}(f::Function, size)` constructor as well?

StefanKarpinski commented on base/channels.jl:
```diff
@@ -105,7 +105,7 @@ true
 Other constructors:
-* `Channel{T}(func::Function, sz=0)`
+* `Channel{T}(func::Function, size)`
 * `Channel{T}(func::Function; csize=0, taskref=nothing)`
```
Is it too late to change these keyword APIs? They're pretty bad. `csize` would be much better spelled `size`—yes, it's the "channel size" but what else would it be the size of? It's the `Channel` constructor. The `taskref` argument is an ok name, but man, this is a really strange API. Is there anywhere else we use a keyword argument to let someone pass in a `Ref` which we populate with an optional output?

NHDaly replied:
> Is it too late to change these keyword APIs? They're pretty bad.

Yeah I completely agree. I was under the impression that we can't change the API for these because of the compatibility promise for julia 1.0? I guess we could at least deprecate the old interface, maybe? Or am I not understanding how constrained we are here? If we can change them, i'm definitely on board!

I think Channels are going to be an important part of our cooperative multithreading story, so it'd be nice to get this API right.

NHDaly replied:
And actually, I want to also open another PR that adds maybe a `spawn=` kwarg that when `true` creates the new Task for the provided function as a multithreading task with `sticky=false`. Something like `Channel(spawn=true) do ch ; ... end`?

I was planning on opening a separate PR for that, just wanted to leave this as a heads-up, while we're thinking about this.

JeffBezanson replied:
Totally agree. `csize` and `ctype` should go. `Channel(ctype = Int)` should obviously be `Channel{Int}()`. `taskref` is weird but at least is solves a real problem (we want the constructor to return a channel, but you might want to get the task as well). The `spawn` keyword arg would be OK too.

StefanKarpinski replied:
Could the task just be accessible as a field of the Channel object?

NHDaly replied:
> Totally agree. `csize` and `ctype` should go.

So this is something we can still do? Backwards compatibility isn't a concern? +1 to removing those

JeffBezanson replied:
We should do a quick grep over all packages to gauge how often `csize` and `ctype` have been used. But `ctype` in particular is so non-idiomatic that I'd be comfortable removing it.

simonbyrne replied:
We can also define multiple keywords and note in the docs that some will be deprecated in the next release. There was a fair few of those in 0.7

NHDaly replied:
> We should do a quick grep over all packages to gauge how often `csize` and `ctype` have been used.

There seems to be 23 uses of `csize` across 9 packages + julia base:
https://pkg.julialang.org/docs/search?q=%5Cbcsize%5Cs%2a%3D

And 99 uses of `ctype` across 34 packages + julia base (if i counted that right):
https://pkg.julialang.org/docs/search?q=%5Cbctype%5Cs%2a%3D

NHDaly replied:
From Triage call:
 - change to the new API, remove the old API from the docs, and describe it as "deprecated" in the News.
 - And maybe start sending PRs to packages to switch to the new API?

NHDaly replied:
Okay this is done. (The changes to julia, not sending PRs to packages)

NHDaly committed:
Make Channel constructors interface more Julian

"Soft" deprecation of the old `Channel(f::Function; csize=0, ctype=Any)`
style in favor of the new `Channel{T=Any}(f::Function, size=0)`
interface.

Removed the old interface from the docs, so that it is not discoverable,
and included a `!!! compat` note that the new interface is new in Julia
1.3, and the old interface should not be used.

NHDaly committed:
Added new Channel constructor to news

NHDaly committed:
Fix test/channels.jl to use newer Channel constructor

NHDaly commented:
Okay, this should be good to go! :) Thanks again for the review.

I think I will open one more PR for adding the `spawn=false` kwarg to the channels so that it can get a more thorough, independent review. Does that sound good? Thanks!

JeffBezanson commented:
Looks good. I think this is missing a `Channel(::Function, size)` constructor (i.e. with no `{T}`)?

NHDaly commented:
Aye, so it is. :) Thanks good catch!

(I was about to fix it by pushing `Channel(func::Function, args...; kwargs...) = Channel{Any}(func, args...; kwargs...)`, but then i realized i could just have a single `T=Any` constructor defined as `Channel(args...; kwargs...) = Channel{Any}( args...; kwargs...)` to cover both the function and non-function cases. Do you think that's valuable? Or is it less clear? Thanks!)

JeffBezanson commented:
Either seems fine to me.

NHDaly commented:
Ah, actually, i think there's an ambiguity with the `ctype=` constructors, so i may have to list the args explicitly anyway. :'(

NHDaly committed:
Add Channel(f, size) = Channel{Any}(...) for default T for task constructor

NHDaly commented:
>Looks good. I think this is missing a `Channel(::Function, size)` constructor (i.e. with no `{T}`)?

Okay thanks! :) Done!

NHDaly committed:
Merge branch 'master' into Channel-constructors

JeffBezanson commented on base/channels.jl:
```diff
@@ -8,9 +8,9 @@ Representation of a channel passing objects of type `T`.
 
 """
-    Channel{T}(sz::Int=0)
+    Channel{T=Any}(size::Int=0)
```
Wow... maybe we should actually implement this syntax for defining constructors with default type parameter values? Seems quite possible.

NHDaly replied:
> Wow... maybe we should actually implement this syntax for defining constructors with default type parameter values? Seems quite possible.

:D Yeah that makes sense to me! It seems pretty readable, and has an intuitive meaning, i think.
